File: csharp/Wjybxx.Dson.Codec/src/Codecs/CollectionCodec.cs
#region LICENSE

// Copyright 2023-2024 wjybxx(845740757@qq.com)
// 
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
// 
//     http://www.apache.org/licenses/LICENSE-2.0
// 
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#endregion

using System;
using System.Collections.Generic;
using Wjybxx.Commons.Collections;

namespace Wjybxx.Dson.Codec.Codecs
{
/// <summary>
/// 集合默认编解码器
/// </summary>
/// <typeparam name="T"></typeparam>
public class CollectionCodec<T> : IDsonCodec<ICollection<T>>
{
    private readonly Type encoderType;
    private readonly Func<ICollection<T>>? factory;
    private readonly FactoryKind factoryKind; // 处理默认情况

    /// <summary>
    /// 动态构建Codec时会被调用
    /// </summary>
    /// <param name="encoderType"></param>
    /// <param name="factory"></param>
    public CollectionCodec(Type encoderType, Func<ICollection<T>>? factory = null) {
        this.encoderType = encoderType;
        this.factory = factory;
        if (factory == null) {
            this.factoryKind = ComputeFactoryKind(encoderType);
        }
    }

    private static FactoryKind ComputeFactoryKind(Type typeInfo) {
        if (typeInfo == typeof(IGenericSet<T>)
            || typeInfo == typeof(LinkedHashSet<T>)) {
            return FactoryKind.LinkedHashSet;
        }
        // Unity没有IReadOnlySet
        if (typeInfo == typeof(ISet<T>)
            || typeInfo == typeof(HashSet<T>)) {
            return FactoryKind.HashSet;
        }
        // Dequeue
        if (typeInfo == typeof(MultiChunkDeque<T>)) {
            return FactoryKind.MultiChunkDequeue;
        }
        if (typeInfo == typeof(IDeque<T>) || typeInfo == typeof(ArrayDeque<>)) {
            return FactoryKind.ArrayDequeue;
        }
        return FactoryKind.Unknown;
    }

    private enum FactoryKind
    {
        Unknown,
        HashSet,
        LinkedHashSet,
        ArrayDequeue,
        MultiChunkDequeue,
    }

    public Type GetEncoderType() => encoderType;

    private ICollection<T> NewCollection(Func<object>? userFactory, int count) {
        if (userFactory != null) return (ICollection<T>)userFactory();
        if (factory != null) return factory();
        return factoryKind switch
        {
            FactoryKind.HashSet => new HashSet<T>(count),
            FactoryKind.LinkedHashSet => new LinkedHashSet<T>(count),
            FactoryKind.ArrayDequeue => new ArrayDeque<T>(),
            FactoryKind.MultiChunkDequeue => new MultiChunkDeque<T>(),
            _ => new List<T>(count)
        };
    }

    private static ICollection<T> ToImmutable(Type declaredType, ICollection<T> result) {
        if (declaredType.IsInterface) {
            if (DsonConverterUtils.IsSet(declaredType)) {
                return ImmutableSet<T>.CreateRange(result);
            }
            if (DsonConverterUtils.IsList(declaredType)) {
                return ImmutableList<T>.CreateRange(result);
            }
        }
        if (declaredType.IsGenericType) {
            if (declaredType.GetGenericTypeDefinition() == typeof(ImmutableSet<>)) {
                return ImmutableSet<T>.CreateRange(result);
            }
            if (declaredType.GetGenericTypeDefinition() == typeof(ImmutableList<>)) {
                return ImmutableList<T>.CreateRange(result);
            }
        }
        return result;
    }

    public void WriteObject(IDsonObjectWriter writer, ICollection<T> inst, Type declaredType, SerializeFeatures features) {
        SerializeFeatures selfFeatures = features.ErasureElementFeatures();
        SerializeFeatures elementFeatures = features.GetElementFeatures();
        // T就是声明类型
        DsonCodecImpl<T> elementCodec = writer.GetInlinableCodec<T>();
        if (elementCodec != null) {
            Type elementType = typeof(T);
            writer.WriteStartArray(encoderType, declaredType, selfFeatures, inst.Count);
            foreach (T value in inst) {
                elementCodec.WriteObject(writer, in value, elementType, elementFeatures);
            }
            writer.WriteEndArray();
        } else {
            writer.WriteStartArray(encoderType, declaredType, selfFeatures, inst.Count);
            foreach (T value in inst) {
                writer.WriteObject(in value, elementFeatures);
            }
            writer.WriteEndArray();
        }
    }

    public ICollection<T> ReadObject(IDsonObjectReader reader, Type declaredType, DeserializeFeatures features, Func<object>? factory = null) {
        DeserializeFeatures selfFeatures = features.ErasureElementFeatures();
        DeserializeFeatures elementFeatures = features.GetElementFeatures();
        //
        int count = reader.ReadStartArray(encoderType).count;
        ICollection<T> result = NewCollection(factory, count);
        // T就是声明类型
        DsonCodecImpl<T> elementCodec = reader.GetInlinableCodec<T>();
        if (elementCodec != null) {
            Type elementType = typeof(T);
            while (reader.ReadDsonType() != DsonType.EndOfObject) {
                T value = elementCodec.ReadObject(reader, elementType, elementFeatures);
                result.Add(value);
            }
        } else {
            while (reader.ReadDsonType() != DsonType.EndOfObject) {
                T value = reader.ReadObject<T>(elementFeatures);
                result.Add(value);
            }
        }
        reader.ReadEndArray();

        // 处理默认的不可变集合
        if (declaredType.IsGenericType) {
            if (declaredType.GetGenericTypeDefinition() == typeof(ImmutableList<>)) {
                return result.ToImmutableList2();
            }
            if (declaredType.GetGenericTypeDefinition() == typeof(ImmutableSet<>)) {
                return result.ToImmutableSet2();
            }
        }
        return DsonCodecHelper.IsReadAsImmutable(features, reader)
            ? ToImmutable(declaredType, result)
            : result;
    }
}
}

File: java/Dson-Codec/src/main/java/cn/wjybxx/dsoncodec/codecs/CollectionCodec.java
/*
 * Copyright 2023-2024 wjybxx(845740757@qq.com)
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package cn.wjybxx.dsoncodec.codecs;

import cn.wjybxx.base.CollectionUtils;
import cn.wjybxx.dson.DsonType;
import cn.wjybxx.dson.text.ObjectStyle;
import cn.wjybxx.dsoncodec.*;
import cn.wjybxx.dsoncodec.annotations.DsonCodecScanIgnore;

import javax.annotation.Nonnull;
import java.util.*;
import java.util.function.Supplier;

/**
 * 集合解码器，动态构造
 *
 * @author wjybxx
 * date 2023/4/4
 */
@DsonCodecScanIgnore
public class CollectionCodec<E> implements DsonCodec<Collection<E>> {

    protected final TypeInfo encoderType;
    protected final Supplier<? extends Collection<E>> factory;
    private final FactoryKind factoryKind;

    public CollectionCodec(TypeInfo encoderType) {
        this(encoderType, null);
    }

    @SuppressWarnings("unchecked")
    public CollectionCodec(TypeInfo encoderType, Supplier<? extends Collection<E>> factory) {
        if (encoderType.genericArgs.size() != 1) {
            throw new IllegalArgumentException("encoderType.genericArgs.size() != 1");
        }
        if (factory == null) {
            factory = DsonConverterUtils.tryNoArgConstructorToSupplier((Class<? extends Collection<E>>) encoderType.rawType);
        }
        this.encoderType = encoderType;
        this.factory = factory;
        this.factoryKind = factory == null ? computeFactoryKind(encoderType) : FactoryKind.Unknown;
    }

    private static FactoryKind computeFactoryKind(TypeInfo typeInfo) {
        Class<?> clazz = typeInfo.rawType;
        if (clazz == EnumSet.class && typeInfo.genericArgs.get(0).isEnum()) {
            return FactoryKind.EnumSet; // 考虑被擦除的情况
        }
        if (Set.class.isAssignableFrom(clazz)) {
            return FactoryKind.LinkedHashSet;
        }
        if (Deque.class.isAssignableFrom(clazz)) {
            return FactoryKind.ArrayDeque;
        }
        return FactoryKind.Unknown;
    }

    private enum FactoryKind {
        Unknown,
        EnumSet,
        LinkedHashSet,
        ArrayDeque,
    }

    @Nonnull
    @Override
    public TypeInfo getEncoderType() {
        return encoderType;
    }

    /** {@link #encoderType}一定是用户declaredType的子类型，因此创建实例时不依赖declaredType */
    @SuppressWarnings({"unchecked", "rawtypes"})
    protected Collection<E> newCollection() {
        if (factory != null) return factory.get();
        return switch (factoryKind) {
            case EnumSet -> {
                TypeInfo elementTypeInfo = encoderType.genericArgs.get(0);
                yield EnumSet.noneOf((Class) elementTypeInfo.rawType);
            }
            case LinkedHashSet -> new LinkedHashSet<>();
            case ArrayDeque -> new ArrayDeque<>();
            default -> new ArrayList<>();
        };
    }

    protected Collection<E> toImmutable(Collection<E> result) {
        if (result instanceof LinkedHashSet<E> linkedHashSet) {
            return Collections.unmodifiableSet(linkedHashSet);
        }
        if (result instanceof EnumSet<?>) {
            Set<? extends E> enumSet = (Set<? extends E>) result;
            return Collections.unmodifiableSet(enumSet);
        }
        if (result instanceof Set<E> set) {
            return CollectionUtils.toImmutableLinkedHashSet(set);
        }
        return List.copyOf(result);
    }

    @Override
    public void writeObject(DsonObjectWriter writer, Collection<E> inst, TypeInfo declaredType, ObjectStyle style) {
        TypeInfo elementTypeInfo = encoderType.genericArgs.get(0);

        for (E e : inst) {
            writer.writeObject(null, e, elementTypeInfo, null);
        }
    }

    @Override
    public Collection<E> readObject(DsonObjectReader reader, Supplier<? extends Collection<E>> factory) {
        TypeInfo elementTypeInfo = encoderType.genericArgs.get(0);

        Collection<E> result = factory != null ? factory.get() : newCollection();
        while (reader.readDsonType() != DsonType.END_OF_OBJECT) {
            result.add(reader.readObject(null, elementTypeInfo));
        }
        return reader.options().readAsImmutable ? toImmutable(result) : result;
    }

}